packages/x-data-grid/src/components/GridRow.tsx

```diff
@@ -285,6 +285,11 @@ const GridRow = forwardRef<HTMLDivElement, GridRowProps>(function GridRow(props,
     return rowStyle;
   }, [isNotVisible, rowHeight, styleProp, heightEntry, rootProps.rowSpacingType]);
 
+  // HACK: Sometimes, the rowNode has already been removed from the state but the row is still rendered.
+  if (!rowNode) {
+    return null;
+  }
```

Comment:
This is the state management issues again, the derived state is computed asynchronously so the `rows` model is already updated while the rendered rows (`gridVisibleRowsSelector`) are lagging behind.